tidy3d/plugins/design/design.py

```diff
@@ -102,78 +144,146 @@ def run(self, function: Callable, **kwargs) -> Result:
+        """Combine pre and post into one big function, stitching tidy3d calls between if needed."""
+
+        def fn_combined(args_list):
+            sim_dict = {idx: fn_pre(**arg_list) for idx, arg_list in enumerate(args_list)}
+            data = self.fn_mid(sim_dict)
+            post_out = [fn_post(val[1]) for val in data.items()]
+            return post_out
+
+        return fn_combined
+
+    def fn_mid(self, pre_out: Any) -> Any:
+        """A function of the output of ``fn_pre`` that gives the input to ``fn_post``."""
+
+        # Convert list of sim inputs to dict of dict before passing through checks
+        was_list = False
+        if all(isinstance(sim, List) for sim in pre_out.values()) and any(
+            isinstance(sim, Simulation) for sim in pre_out[0]
+        ):
+            pre_out = {
+                idx: dict(enumerate(sim_list)) for idx, sim_list in enumerate(pre_out.values())
+            }
+            was_list = True
+
+        # Handle most common case where fn_combined builds a dict of sims
+        if all(isinstance(sim, Simulation) for sim in pre_out.values()):
+            named_sims = {next(self._name_generator): sim for sim in pre_out.values()}
+            self._task_names.extend(list(named_sims.keys()))
+            data = web.Batch(simulations=named_sims, folder_name=self.folder_name).run(
+                path_dir=self.path_dir
+            )
+
+        # Can "index" dict here because fn_combined uses idx as the key
+        elif all(isinstance(sim, Dict) for sim in pre_out.values()) and any(
+            isinstance(sim, Simulation) for sim in pre_out[0].values()
+        ):
+            # Flatten dict of dicts whilst storing combined keys for later
+            flattened_sims = {}
+            original_structure = {}
+            dict_keys = []
+            for dict_idx, sub_dict in pre_out.items():
+                sub_structure = {}
+                for sub_dict_idx, sim_like in sub_dict.items():
+                    sub_dict_idx = str(sub_dict_idx)  # Needs to be string for restoration later
+                    if isinstance(sim_like, Simulation):
+                        task_name = f"{dict_idx}_{sub_dict_idx}"
+                        flattened_sims[task_name] = sim_like
+                        sub_structure[sub_dict_idx] = None
+
+                        # Add user specified keys to the naming if a dict was supplied
+                        if not was_list:
+                            dict_keys.append(sub_dict_idx)
+                        else:
+                            dict_keys.append(None)
+                    else:
+                        sub_structure[sub_dict_idx] = sim_like
+
+                original_structure[dict_idx] = sub_structure
+
+            # Run sims with flattened dict
+            named_sims = {}
+            for sim, key_name in zip(flattened_sims.values(), dict_keys):
+                self._dict_name = key_name
```

Comment:
if we get rid of the generator, we can also remove the need for this? and also the `dict_keys`?